In [1]:
import pandas as pd
import numpy as np
import warnings
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
warnings.filterwarnings('ignore')

In [2]:
churnData = pd.read_csv('Customer-Churn.csv')
churnData

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
2       Male              0      No         No       2          Yes   
3       Male              0      No         No      45           No   
4     Female              0      No         No       2          Yes   
...      ...            ...     ...        ...     ...          ...   
7038    Male              0     Yes        Yes      24          Yes   
7039  Female              0     Yes        Yes      72          Yes   
7040  Female              0     Yes        Yes      11           No   
7041    Male              1     Yes         No       4          Yes   
7042    Male              0      No         No      66          Yes   

     OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV  \
0                No          Yes               No          No          No   
1               Yes           No              Yes          No          No   
2               Yes          Yes               No          No          No   
3               Yes           No              Yes         Yes          No   
4                No           No               No          No          No   
...             ...          ...              ...         ...         ...   
7038            Yes           No              Yes         Yes         Yes   
7039             No          Yes              Yes          No         Yes   
7040            Yes           No               No          No          No   
7041             No           No               No          No          No   
7042            Yes           No              Yes         Yes         Yes   

     StreamingMovies        Contract  MonthlyCharges TotalCharges Churn  
0                 No  Month-to-month           29.85        29.85    No  
1                 No        One year           56.95       1889.5    No  
2                 No  Month-to-month           53.85       108.15   Yes  
3                 No        One year           42.30      1840.75    No  
4                 No  Month-to-month           70.70       151.65   Yes  
...              ...             ...             ...          ...   ...  
7038             Yes        One year           84.80       1990.5    No  
7039             Yes        One year          103.20       7362.9    No  
7040              No  Month-to-month           29.60       346.45    No  
7041              No  Month-to-month           74.40        306.6   Yes  
7042             Yes        Two year          105.65       6844.5    No  

[7043 rows x 16 columns]

In [3]:
#Check the datatypes of all the columns in the data
churnData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   object 
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   object 
 3   Dependents        7043 non-null   object 
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   object 
 6   OnlineSecurity    7043 non-null   object 
 7   OnlineBackup      7043 non-null   object 
 8   DeviceProtection  7043 non-null   object 
 9   TechSupport       7043 non-null   object 
 10  StreamingTV       7043 non-null   object 
 11  StreamingMovies   7043 non-null   object 
 12  Contract          7043 non-null   object 
 13  MonthlyCharges    7043 non-null   float64
 14  TotalCharges      7043 non-null   object 
 15  Churn             7043 non-null   object 
dtypes: float64(1), int64(2), object(13)
memory

In [4]:
churnData['TotalCharges'] = churnData['TotalCharges'].str.replace(" ", "") #it was not being converted to number without rremoving the space from a category

In [5]:
#Converting the type of TotalCharges to int:
churnData['TotalCharges'] = pd.to_numeric(churnData['TotalCharges'])


In [6]:
churnData.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 16 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            7043 non-null   object 
 1   SeniorCitizen     7043 non-null   int64  
 2   Partner           7043 non-null   object 
 3   Dependents        7043 non-null   object 
 4   tenure            7043 non-null   int64  
 5   PhoneService      7043 non-null   object 
 6   OnlineSecurity    7043 non-null   object 
 7   OnlineBackup      7043 non-null   object 
 8   DeviceProtection  7043 non-null   object 
 9   TechSupport       7043 non-null   object 
 10  StreamingTV       7043 non-null   object 
 11  StreamingMovies   7043 non-null   object 
 12  Contract          7043 non-null   object 
 13  MonthlyCharges    7043 non-null   float64
 14  TotalCharges      7032 non-null   float64
 15  Churn             7043 non-null   object 
dtypes: float64(2), int64(2), object(12)
memory

In [7]:
#Check for null values in the dataframe. Replace the null values.
churnData.isna().sum()

gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

In [8]:
#Dropping the nulls:
churnData.dropna()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0     Female              0     Yes         No       1           No   
1       Male              0      No         No      34          Yes   
2       Male              0      No         No       2          Yes   
3       Male              0      No         No      45           No   
4     Female              0      No         No       2          Yes   
...      ...            ...     ...        ...     ...          ...   
7038    Male              0     Yes        Yes      24          Yes   
7039  Female              0     Yes        Yes      72          Yes   
7040  Female              0     Yes        Yes      11           No   
7041    Male              1     Yes         No       4          Yes   
7042    Male              0      No         No      66          Yes   

     OnlineSecurity OnlineBackup DeviceProtection TechSupport StreamingTV  \
0                No          Yes               No          No          No   
1               Yes           No              Yes          No          No   
2               Yes          Yes               No          No          No   
3               Yes           No              Yes         Yes          No   
4                No           No               No          No          No   
...             ...          ...              ...         ...         ...   
7038            Yes           No              Yes         Yes         Yes   
7039             No          Yes              Yes          No         Yes   
7040            Yes           No               No          No          No   
7041             No           No               No          No          No   
7042            Yes           No              Yes         Yes         Yes   

     StreamingMovies        Contract  MonthlyCharges  TotalCharges Churn  
0                 No  Month-to-month           29.85         29.85    No  
1                 No        One year           56.95       1889.50    No  
2                 No  Month-to-month           53.85        108.15   Yes  
3                 No        One year           42.30       1840.75    No  
4                 No  Month-to-month           70.70        151.65   Yes  
...              ...             ...             ...           ...   ...  
7038             Yes        One year           84.80       1990.50    No  
7039             Yes        One year          103.20       7362.90    No  
7040              No  Month-to-month           29.60        346.45    No  
7041              No  Month-to-month           74.40        306.60   Yes  
7042             Yes        Two year          105.65       6844.50    No  

[7032 rows x 16 columns]

In [9]:
churnData['Churn'].value_counts()

No     5174
Yes    1869
Name: Churn, dtype: int64

We have an imbalanced target

In [ ]:
#Encoding categoricals:


In [10]:
#Scale the features either by using normalizer or a standard scaler:

X = churnData[['tenure', 'SeniorCitizen', 'MonthlyCharges', 'TotalCharges']]
y = churnData['Churn']
X

tenure  SeniorCitizen  MonthlyCharges  TotalCharges
0          1              0           29.85         29.85
1         34              0           56.95       1889.50
2          2              0           53.85        108.15
3         45              0           42.30       1840.75
4          2              0           70.70        151.65
...      ...            ...             ...           ...
7038      24              0           84.80       1990.50
7039      72              0          103.20       7362.90
7040      11              0           29.60        346.45
7041       4              1           74.40        306.60
7042      66              0          105.65       6844.50

[7043 rows x 4 columns]

In [11]:
#Scaling all the discrete and continous columns:
from sklearn.preprocessing import MinMaxScaler
scaler = MinMaxScaler()
scaled = scaler.fit_transform(X)

In [12]:
X = pd.DataFrame(scaled,columns=X.columns)

In [13]:
X.isna().sum()

tenure             0
SeniorCitizen      0
MonthlyCharges     0
TotalCharges      11
dtype: int64

In [14]:
X =X.replace(np. nan,0) 

In [15]:
#Split the data into a training set and a test set.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=100)

In [16]:
#Fit a logistic regression model on the training data.
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import classification_report
from sklearn.metrics import cohen_kappa_score 
classification = LogisticRegression(random_state=0).fit(X_train, y_train)

In [17]:
predictions = classification.predict(X_test)

In [18]:
predictions

array(['No', 'No', 'No', ..., 'No', 'No', 'No'], dtype=object)

In [19]:
#Accuracy
classification.score(X_test, y_test)

0.7808802650260294

In [20]:
from sklearn.metrics import confusion_matrix
confusion_matrix(y_test, predictions)

array([[1397,  150],
       [ 313,  253]], dtype=int64)

The accuracy is high. However, that could be because the model is good at predicting the dominant category (No). Let's try to balance the data and see if it changes

**Using SMOTE oversampling technique to handle data imbalance**

In [21]:
from imblearn.over_sampling import SMOTE
smote = SMOTE()
#oversampling
X_sm, y_sm = smote.fit_resample(X, y)
y_sm.value_counts()

No     5174
Yes    5174
Name: Churn, dtype: int64

**Use logistic regression to fit the model and compute the accuracy of the model.**

In [22]:
#applying logistic regression to predict the target
X_train, X_test, y_train, y_test = train_test_split(X_sm, y_sm, test_size=0.3, random_state=100)
classification = LogisticRegression(random_state=0, multi_class='ovr').fit(X_train, y_train)
predictions = classification.predict(X_test)
#Accuracy of the predictions
classification.score(X_test, y_test)

0.743317230273752

In [23]:
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

          No       0.74      0.75      0.74      1557
         Yes       0.74      0.74      0.74      1548

    accuracy                           0.74      3105
   macro avg       0.74      0.74      0.74      3105
weighted avg       0.74      0.74      0.74      3105



The classification score reduced compared to the time that we had imbalanced data. However, this model is more reliable.

**Use decision tree classifier to fit the model and compute the accuracy of the model.**

In [25]:
X_train, X_test, y_train, y_test = train_test_split(X_sm, y_sm, test_size=0.3, random_state=100)
from sklearn.tree import DecisionTreeClassifier
model = DecisionTreeClassifier(criterion='entropy', max_depth=None, random_state=100)
model.fit(X_train, y_train)
model.score(X_test, y_test)

0.7594202898550725

Summary: Using the decision tree, the accuracy increased from 0.74 to 0.76.

**Using TomekLinks to reduce the noise of the majority class:**

**Use logistic regression to fit the model and compute the accuracy of the model.**

In [27]:

from imblearn.under_sampling import TomekLinks

tl = TomekLinks(sampling_strategy='majority')
X_tl, y_tl = tl.fit_resample(X, y)
y_tl.value_counts()

No     4651
Yes    1869
Name: Churn, dtype: int64

In [28]:
#Using logistic regression:
X_train, X_test, y_train, y_test = train_test_split(X_tl, y_tl, test_size=0.3, random_state=100)
classification = LogisticRegression(random_state=0, multi_class='ovr').fit(X_train, y_train)
predictions = classification.predict(X_test)
#Calculating the accuracy
classification.score(X_test, y_test)

0.7955010224948875

In [29]:
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions))

              precision    recall  f1-score   support

          No       0.82      0.91      0.86      1388
         Yes       0.70      0.52      0.60       568

    accuracy                           0.80      1956
   macro avg       0.76      0.71      0.73      1956
weighted avg       0.79      0.80      0.79      1956



**Use decision tree classifier to fit the model and compute the accuracy of the model.**

In [30]:
from sklearn.tree import DecisionTreeClassifier
X_train, X_test, y_train, y_test = train_test_split(X_tl, y_tl, test_size=0.3, random_state=100)
model = DecisionTreeClassifier(criterion='entropy', max_depth=None, random_state=100)
model.fit(X_train, y_train)
model.score(X_test, y_test)

0.7402862985685071

Summary: Using Tomeklinks and decision tree, the accuracy reduced from 0.79 to 0.74 compared to logistic regression and tomeklinks.

**Decision tree with imbalanced target**

In [31]:
#Split the data into a training set and a test set.
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=100)
model = DecisionTreeClassifier(criterion='entropy', max_depth=None, random_state=100)
model.fit(X_train, y_train)
model.score(X_test, y_test)

0.722669190724089

Summary: The accuracy reduced in imbalanced target when decision tree modeling was used compared to logistic regression (from 0.79 to 0.72)